File: java/omnia/data/structure/immutable/ArrayIterator.java
package omnia.data.structure.immutable;

import java.util.Iterator;

final class ArrayIterator<E> implements Iterator<E> {
  private final E[] elements;
  private int i = 0;

  ArrayIterator(E[] elements) {
    this.elements = elements;
  }

  @Override
  public boolean hasNext() {
    return i < elements.length;
  }

  @Override
  public E next() {
    return elements[i++];
  }
}

File: omnia/src/commonMain/omnia/data/iterate/ArrayIterator.kt
package omnia.data.iterate

import kotlin.collections.Iterator

/** A simple read-only iterator that iterates over the elements of an array in sequential order. */
class ArrayIterator<E>(private val elements: Array<E>) : Iterator<E> {

  private var i = 0

  override fun hasNext(): Boolean {
    return i < elements.size
  }

  override fun next(): E {
    if (i < elements.size) {
      return elements[i++]
    }
    throw NoSuchElementException("ArrayIterator reached last index in array: $i")
  }
}